src/main/java/com/google/sps/servlets/ImageServlet.java

```diff
@@ -0,0 +1,145 @@
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.protobuf.ByteString;
+import com.google.gson.Gson;
+import java.io.ByteArrayOutputStream;
+import java.net.URL;
+import java.net.MalformedURLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/user-image")
+public class ImageServlet extends HttpServlet {
+  
+  // TODO: Handle keywords in a datastore, and handle the adding and replacing of keywords within
+  // the datastore in a separate class in the data package.
+  private static final int MAX_NUM_KEYWORDS = 10;
+
+  /**
+   * Writes an upload URL for file uploads to the servlet.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstore = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstore.createUploadUrl("/user-image");
+    response.getWriter().println(uploadUrl);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PrintWriter writer = response.getWriter();
+    BlobKey blobKey = getBlobKey(request, "image");
+    if (blobKey == null) {
+      response.sendRedirect("/index.html");
+      return;
+    }
+    byte[] blobBytes = getBlobBytes(blobKey);
+    List<EntityAnnotation> annotations = getImageLabels(blobBytes);
+   
+    // Create an array of labels for printing the appropriate JSON.
+    List<String> imageLabels = new ArrayList<>();;
+    for (int i = 0; i < imageLabels.size() && i < MAX_NUM_KEYWORDS; i++) {
+      imageLabels.add(annotations.get(i).getDescription());
+    }
+    Gson gson = new Gson();
+    String json = gson.toJson(imageLabels);
+    writer.println(json);
+    response.sendRedirect("/main.html");
+  }
+
+  /**
+   * Returns the BlobKey that points to the file uploaded by the user, or null if the user didn't
+   * upload a file.
+   */
+  private BlobKey getBlobKey(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+    System.out.println(blobKey.getKeyString());
+
+    // User submitted form without selecting a file, so the BlobKey is empty (live).
```

Comment:
[Optional] Maybe say "(prod server)" instead of "(live)" as I think it makes it a bit clearer that this is due to a difference between the dev and prod servers.